New theme for luigi visualiser

stephenpascoe opened the pull request:
This is a new theme for the visualiser accessible through .../index.lte.html.  The theme takes the d3 theme and adds jQuery DataTables for displaying a filterable table of tasks and AdminLTE for styling.

Key features are:

* Table paging
* Search filter by any text in the table
* Info boxes displaying count of task status (done, running, etc.) are selectable to filter the table
* Sidebar displaying task names with counts are selectable to filter the table
* Exception reports are accessible via expandable rows

I can imagine improving the layout of the "Parameters" column and making more use of the expandable rows but we are already finding this very useful in production.

With 3 separate versions of index.html and several new dependencies I think it would be worth refactoring the /static structure and finding a neater way of selecting themes.  If you agree I'll take a look at it.

Tarrasch commented:
Pics or it didn't happen :p

stephenpascoe commented:
Screenshot with one task name selected.

![luigi_lte_theme](https://cloud.githubusercontent.com/assets/677427/8830834/61b15f58-3096-11e5-8404-01868cdbb5f2.png)

graingert commented on luigi/static/visualiser/index.lte.html:
```diff
@@ -0,0 +1,481 @@
+
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+
+        <style type="text/css">
```
can these styles be static files?

stephenpascoe replied:
I just copied the inline style from index.d3.html.  I can definitely factor them into a separate file.  It would be good to do a more general refactor of the 3 themes at the same time as they are rather mixed up at the moment.

Tarrasch commented:
Does this patch come with some documentation of how to set it up and stuff? Or are you just supposed to browse to `static/visualizer/index.lte.html`?

erikbern commented:
This is beautiful

stephenpascoe commented:
@erikbern thanks! 
@Tarrasch I'm working on configuration now to switch themes.  No setup should be needed, just navigate to index.lte.html

stephenpascoe commented:
I have now made a more general refactoring of the visualiser files to put each theme in its own directory.  The theme can be set with the "visualizer_theme" option as either "d3", "svg" or "lte".

I've kept the "visualization_graph" option for backward compatibility but it is a bit confusing having both of these options.  I'm wondering whether it is acceptable to have one option named "visualizer_theme" instead, particularly as this option is not in the docs yet.  Once I get feedback on that I can write a section in the configuration docs.

erikbern commented:
wait we have three themes now? can we deprecate one or two of them and change the default?

stephenpascoe commented:
Sorry, I'm conflating 2 things.  I've written the new theme to use the d3 visualisation.  My understanding is that some people want to continue to use the svg visualisation, therefore I've merged the concepts into 3 themes:
 1. "svg": traditional style with svg visualisation
 2. "d3": traditional style with d3 visualisation
 3. "lte": new style with d3 visualisation

I think it's too much work to make the visualisation independent from the style and I'm using the d3 visualisation in my environment.  So, it's a question of what people want.  We could remove no. 2?

erikbern commented:
I think we should remove #1 and #2, your visualizer looks much better! I'm sure people will disagree though

stephenpascoe commented:
OK, I'm happy with that.  My only hesitation is that I haven't  tried the d3 visualisation with very large dependency trees.

I'll make mine the default and retain the original as an option.

gpoulin commented:
I agree with @erikbern on this one. The frontend is just a monitoring tool. I don't having many themes bring so much, but it complexifies any change to the frontend.

interskh commented:
We tried D3 one a while back. It doesn't work well with large dependency trees unfortunately :(

But I like this new theme!

Tarrasch commented:
@interskh oh dang, I hoped D3 would be more efficient :-/

stephenpascoe commented:
I've had a look at scalability of the D3 visualisation.

 - There is a bug in zoom/pan which resets the view periodically.  This makes navigating trees a pain if it doesn't fit in the window.
 - Up to a few 100 leaf nodes D3 is passable.  Draw speed is ok.  All nodes at a given depth are rendered at the same x-coord which makes wide trees tall and thin.
 - At ~1000 leaf nodes redraw speed is too slow.

It would be useful to know what constitutes a large tree for others.

This PR could be changed to use the old visualisation but supporting both would require refactoring and add complexity.  Right now, I'm finding the D3 vis useful so I'd like to find a way to make it work.  I'll look more closely at the code to consider options.

I'm going to be offline for most of the next 2 weeks so won't get back to this for a while.

interskh commented:
For an example, We have a task depends on a month of hourly task, each of which depends on 2 more tasks. So there are 2k nodes and D3 just refuses to draw anything (or just hang), svg actually renders pretty fast (despite not pretty, but it is still useful) :(

Another thing is that the check for "Show Upstream Dependencies" doesn't seem to work on d3.

tym-oao commented:
I can write this as a separate issue elsewhere if that would be better, but as long as we're griping about d3: I've noticed all the above issues with zoom and large graphs, and plus the "Show Error Trace" button seems to show just an empty lightbox—at least when I use it.

kwent commented:
This is beautiful ! Goob job can't wait to see that merge with themes supports !

erikbern commented:
merge?

Tarrasch commented:
I think we're waiting for @stephenpascoe to come back from vacation and reorganize this PR a bit. I'm also not sure who've actually closely reviewed this.

stephenpascoe commented:
Just to let you know I'm back and have some time to work on this PR this week.  I've sped D3 up a little and have found a way to get both visualisations switchable in the same page.  I hope to update the PR this week.

interskh commented:
:+1:

Tarrasch commented:
:+1: indeed :)

beeva-luisgonzalez commented:
looking good!

stephenpascoe commented:
I have removed the theme logic so there is only one theme (the new one).  All references to "themes" have been removed.

 - You can switch between SVG and D3 visualisations using a switch on the page
 - I changed the D3 layout algorithm to one which draws a little quicker.  ~500 nodes are rendered in ~4s.  The layout is slightly less optimal but looks OK to me.
 - I removed the automatic refresh code.  This caused problems when dragging the graph and was infeasible for medium-sized graphs
 - I fixed display of the inverse dependency graphs with D3

Tarrasch commented:
Cool! Awesome with a switch on the page.

> I removed the automatic refresh code. This caused problems when dragging the graph and was infeasible for medium-sized graphs

Great that you prioritize stability. :)

Tarrasch commented:
@stephenpascoe. By the way, does this new visualizer just deman more on the frond end (your laptop/browser) or does it demand more from luigid as well? Will it make additional API calls?

stephenpascoe commented:
With auto-refresh removed it doesn't make any more API calls than the original.  All table sorting and filtering happens on data loaded on page load.  You have to reload to refresh the data.

Tarrasch commented:
Awesome! :)

Tarrasch commented:
How stable is this to merge you think @stephenpascoe?  I see there's a bug commit fix. Have you used this in production?

stephenpascoe commented:
This is stable IMO.

Tarrasch commented:
Hi @stephenpascoe, I can't get this to work locally. I *suck* at web development. But it seems like it can't find the files `AdminLTE-app.min.js`. Where can I find those files? It's not uploaded from this PR it seems, but they are referenced it seems from `luigi/static/visualiser/index.html`. There are also much more deletion lines over insertion lines. Did you forget to upload those files? :)

![selection_051](https://cloud.githubusercontent.com/assets/294349/9440557/571ba41a-4a70-11e5-80f5-50afe5350c61.png)

stephenpascoe commented:
Ok, .gitignore has been stopping me adding some JS resources and I haven't noticed.  I'll check these again and update.

Tarrasch commented:
Thanks @stephenpascoe :)

stephenpascoe commented:
Hopefully everything is included now.

Tarrasch commented:
Cool. I think it works for me now. But I have some feedback that I think has to be addressed before we can merge this. :)

* It seems like the (`[+]`)-button isn't that useful. It seems to show the error trace for failed tasks. But it doesn't do anything for non-failed tasks. Is that correct?

![selection_052](https://cloud.githubusercontent.com/assets/294349/9441318/b5c5c1ee-4a75-11e5-87e9-64fa54b3f0a6.png)

 * It seems like the filtering for "UPSTREAM FAILURE" doesn't work. See the image below:

![selection_053](https://cloud.githubusercontent.com/assets/294349/9441437/5b9da9d8-4a76-11e5-97c1-a9e2b02ce6f1.png)

I'll continue in a new comment just so I don't lose the content of this post. :)

stephenpascoe commented:
On [+].  Yes, presently it only displays error traces.  Personally, I prefer that to a pop-up.  I can imagine putting more information in the [+] dropdown, maybe reducing the width of the basic table, but I haven't thought deeply about it yet.  I'm open to suggestions on how/whether we use this.

I'll take a look at UPSTREAM_FAILURE.

Tarrasch commented:
I get the bug that I can't scroll down on the page a lot of times. I still don't know how to reproduce this. This means that I can't see all the column families on the left pane.

![selection_055](https://cloud.githubusercontent.com/assets/294349/9441567/36a75d9e-4a77-11e5-983a-1ab6ed24742c.png)

Tarrasch commented:
As for the [+] expansion. Could it be on the right side of the "graph" button like it is in the old visualizer? Clicking it doesn't need to open an alert window, it could just expand the task like you do for [+] but it would only be click-able for tasks where it's meaningful to click it.

Tarrasch commented:
Shouldn't the FAILED tasks be red in the d3 visualizer? What do you think about the clicking behaviour? For the d3 it brings you to that task's visualisation page ...

![selection_056](https://cloud.githubusercontent.com/assets/294349/9441690/e2c171f0-4a77-11e5-91fd-14f43c51d0b3.png)

... but for the svg visualizer it unsuccessfully "tries" to fetch the error message. However it just "darkens" the page and my browser tab freezes, I have to reload the page to get out of it.

![selection_057](https://cloud.githubusercontent.com/assets/294349/9441765/4853e214-4a78-11e5-9143-2bb9b20805db.png)

Tarrasch commented:
Anway, really awesome work on this @stephenpascoe! I hope you can reproduce some of the errors that I see and it would be really good to get them fixed. Then I think this can be merged. :)

stephenpascoe commented:
Thanks for the bug reports.  I'll work on these during EuroSciPy later this week.

Tarrasch commented:
Cool :)

stephenpascoe commented:
@Tarrasch, I think I've fixed everything you brought up here, and a few other things along the way.

Tarrasch commented:
Awesomeness!!!

Do you want to squash the commits? I leave this up to you to decide.

Tarrasch commented:
@daveFNbuck, do you want to review this and/or try out the visualizer on your graphs to make sure it makes sense for you  guys too? I was able to spot some things that wouldn't work for Spotify as you can see in the comments. I billion thanks to @stephenpascoe for fixing them too. :)

stephenpascoe commented:
I think the commits make sense as they are.  They are all atomic fixes and improvements.

Tarrasch commented:
Great, I'll merge in a few days (say Tuesday?) in case somebody else want to test this out.

daveFNbuck commented:
I'll try to check them out tomorrow, but no need to hold things up on my account.

Tarrasch commented:
It's always good to have an extra pair of eyes on things :)

daveFNbuck commented:
I just got it installed and overall I love it. It looks much nicer and is a welcome change to a page I have to look at all the time. I really like how the information at the top of the task list page now gives the high-level view that I'm often looking for when I refresh the page. However, I do have a few issues. 

First, I have a couple thousand pending tasks but only 367 show up as pending on the main page. The worker page shows more than that. Did the internal representation change in a way that requires clearing the pickle?

My done task count is beyond the limit where the server sends them all, so I get a question mark rather than a count on the tasks page.

I've lost the ability to search within done tasks when there are too many to display. This used to trigger a call to the server to fetch tasks that matched the query but no longer does.

The task page seems to be ignoring tasks that were disabled by the scheduler. They're missing when I search for them and don't show up in the disabled task count. I need to be able to see them and re-enable them with an action button.

The action icons need hover text because their functionality is not at all obvious.

Using D3 as the default graph is way too slow and makes the graph view unusable for me.

The SVG graph no longer scales to my window width.

The workers tab did not survive the transition very well. The tasks are fairly hard to read compared to the nice layout of everything else in the new vis. It's a bit jarring.

It's hard to tell when something in the side bar is selected. I got confused a few times as to why my searches were giving such limited results until I realized I had left something selected. This would be more obvious if the highlight color stood out more than black on gray.

Same issue with disabled and upstream disabled, it's not obvious that they're selected because the highlight color blends in with the background.

The hamburger menu item has an extremely unintuitive purpose and I'm not sure why it exists. It might make a little sense if it completely hid the side bar, but it's confusing and unnecessary as-is.

The sidebar should probably go away on the graph and worker tabs, as it doesn't affect them.

The sidebar has two titles but neither one is informative and the link in the top one duplicates the Task List link. They can probably be safely removed.

Tarrasch commented:
That's quite some good feedback! This visualiser is going to be so great! :)

interskh commented:
We forgot to mention priority and resources seem missing from the task list.

stephenpascoe commented:
@daveFNbuck thanks for the great feedback.  I wasn't expecting this to need so much polishing but I'm up for getting it right.

> First, I have a couple thousand pending tasks but only 367 show up as pending on the main page. The worker page shows more than that. Did the internal representation change in a way that requires clearing the pickle?

Are any of these pending tasks upstream failures or disaabled?  I'm counting these independently.  Maybe this is also related to the server limit (see below)?

> My done task count is beyond the limit where the server sends them all, so I get a question mark rather than a count on the tasks page.
> I've lost the ability to search within done tasks when there are too many to display. This used to trigger a call to the server to fetch tasks that matched the query but no longer does.

I wasn't aware that the server has a limit to how many tasks are returned.  I'll have to take a look at how this could be overcome.  Although DataTables can work with server-side queries it works best when all data is retrieved and sorting/filtering happens in the client.  An extra box to "filter server side" might be needed.  Is there a way to tell from the API that not all results have been returned?

> The task page seems to be ignoring tasks that were disabled by the scheduler. They're missing when I search for them and don't show up in the disabled task count. I need to be able to see them and re-enable them with an action button.

I haven't tested with disabled tasks because my workflows typically don't use them.  Providing they are easy to simulate I should be able to get this fixed.

> The action icons need hover text because their functionality is not at all obvious.
Agreed.

> Using D3 as the default graph is way too slow and makes the graph view unusable for me.

We know the D3 visualisation doesn't scale to thousands of tasks but one of the visualisations has to be default.  It's awkward to make it configurable without changing how tornado serves the content.
Would it suffice for the visualiser to be specified in the URL fragment (http:/.../index.html#visType=...)?

> The SVG graph no longer scales to my window width.

OK, I opted for making the DIV scrollable.  I'll take a look at what it did before.

> The workers tab did not survive the transition very well. The tasks are fairly hard to read compared to the nice layout of everything else in the new vis. It's a bit jarring.

I haven't worked on the worker tab at all.  I would rather see this addressed as a future update unless it is completely broken.

> It's hard to tell when something in the side bar is selected. I got confused a few times as to why my searches were giving such limited results until I realized I had left something selected. This would be more obvious if the highlight color stood out more than black on gray.

I agree, some more visual feedback on what the search constraints are would be useful.  I can do that.

> Same issue with disabled and upstream disabled, it's not obvious that they're selected because the highlight color blends in with the background.

I can change the colours but they look clear to me.

> The hamburger menu item has an extremely unintuitive purpose and I'm not sure why it exists. It might make a little sense if it completely hid the side bar, but it's confusing and unnecessary as-is.

This is a bug introduced when I fixed long task lists being inaccessible.  It should completely hide the sidebar.  I will fix.

> The sidebar should probably go away on the graph and worker tabs, as it doesn't affect them.

It could be hidden automatically when you select the graph tab -- providing I can fix the hamburger menu issue above.

> The sidebar has two titles but neither one is informative and the link in the top one duplicates the Task List link. They can probably be safely removed.

The intention is that the "Luigi Task Status" title is part of the top bar, not the sidebar.  I'm agnostic on whether we need the "TASKS" title or what it should say.  Is "Task Family" well understood?  The code uses "Task Name" interchangeably.

erikbern commented:
Very excited about this!

daveFNbuck commented:
> Are any of these pending tasks upstream failures or disabled? I'm counting these independently.

That probably explains it, as all the other counts seemed accurate and jobs had been building up yesterday due to some overnight failures. The previous version placed such tasks both in upstream failed/disabled and pending. It might be confusing to users to change it, but I can see why you did. I think I can get used to either solution. Are upstream failed and upstream disabled all pending?

> Although DataTables can work with server-side queries it works best when all data is retrieved and sorting/filtering happens in the client. An extra box to "filter server side" might be needed.

We do the server-side filtering because the visualizer is very slow when dealing with tens of thousands of tasks. The new version seems snappier, so maybe it can handle a higher limit and remove this obstacle. I can test that today.

> Is there a way to tell from the API that not all results have been returned?

The API call will return a dict containing only the number of tasks, with the format {"num_tasks": 10250}

> I haven't tested with disabled tasks because my workflows typically don't use them. Providing they are easy to simulate I should be able to get this fixed.

You can load a scheduler-disabled task by setting

[scheduler]
disable-num-failures: 1

in your luigi.cfg and then scheduling a job as failed. That's the kind of disabled task that was missing completely from the datatable. I can look a bit into why this was happening today too, as it's probably something weird to do with how scheduler disables are represented internally.

> Would it suffice for the visualiser to be specified in the URL fragment (http:/.../index.html#visType=...)?

I think that would be a bit tricky for my team, as I'd have to tell everyone to start using a new URL. It would also require tweaking all your code to propagate that part of the URL across tabs. Better solutions would be to remember my last choice in a cookie or to accept the default from the server. Or I can just change the relevant line in my fork. I don't know how many other users need a different default.

> I haven't worked on the worker tab at all. I would rather see this addressed as a future update unless it is completely broken.

That's completely reasonable. You might see that pull request from my team soon.

> I can change the colours but they look clear to me.

It might just need to be slightly darker, or I may just need to get more used to the new interface. The highlight just doesn't stand out nearly as much as for the other categories. Maybe this should be a different color, as it's hard for gray to stand out much. This one isn't a huge issue, so you can skip it for now.

> The intention is that the "Luigi Task Status" title is part of the top bar, not the sidebar. I'm agnostic on whether we need the "TASKS" title or what it should say. Is "Task Family" well understood? The code uses "Task Name" interchangeably.

I think that Task Family would be a lot more clear there.

daveFNbuck commented:
I just tested the pending bug again. When I query the API via my slack bot, I see 1799 jobs pending. When I look at the scheduler, I see 279 pending tasks, 33 upstream failures and 12 upstream disables. There's still a good chunk of jobs unaccounted for. I'll dig into this a bit more to see what's happening.

daveFNbuck commented:
After digging into this a bit more, it seems that the visualizer is dropping any PENDING tasks that have resources. It seems to be fine with showing them if they're DISABLED/FAILED/RUNNING, but the PENDING ones are getting dropped somewhere. I don't see the issue with the visualizer not showing scheduler disabled tasks anymore, but maybe this was a coincidence having to do with other features of the task like the resources bug (I do see scheduler disabled tasks with resources on the visualizer now though).

daveFNbuck commented:
DONE tasks have the same problem as pending after increasing the server limit so that they show up in the vis again. It's dropping the ones with resources, plus most of the others.

Tarrasch commented:
Great findings. This seem to be going in the right direction!

@daveFNbuck Haha, there's a luigi slack bot? :)

daveFNbuck commented:
We have luigi commands built into our slack bot, yeah. It can do most API calls. We should probably open source it now that I realize other people might want it.

daveFNbuck commented:
Ok, I ran this through a javascript debugger and found the culprit. In updateTaskCategory, you're setting taskMap[data.name] = category. Unfortunately, data.name is the family name, so all tasks of the same family end up being labeled with the same category. So because a few of my tasks are upstream pending and upstream disabled, everything in the same family with lower priority categories gets filtered out. This happens even though they were never added to the higher priority categories. So stuff just gets randomly removed from the visualizer.

There's also another issue with how you're computing mostImportantCategory. In order for mostImportantCategory to work properly, you need to know whether a task belongs in a higher priority category, but you perform category updates in a different order than the priority. You should probably have a constant to ensure that the priority order and the update order are the same. You don't really need to do this though, as you'll only see conflicts between upstream and regular categories. So you can change the logic to be more explicit about that and only have to worry about updating upstream categories first. You probably also only need to apply these to PENDING, as all other categories will supersede any upstream status.

daveFNbuck commented:
My issues with missing PENDING and DISABLED tasks go away when I replace taskName with taskId in updateTaskCategory. I still can't allow the full list of DONE tasks into the datatable, because that one takes about 10 seconds to load, after which the page freezes for 5 or 10 seconds while the data processes.

stephenpascoe commented:
Now working on these issues in a new branch https://github.com/stephenpascoe/luigi/tree/visualiser.  I'll merge back into this PR when ready.

dymov commented:
Exciting, this will be a superb upgrade to an already awesome product! :)

daveFNbuck commented:
My team and I are very excited about being able to use this. Thanks for working so hard on it, @stephenpascoe!

erikbern commented:
yes very cool. looking forward to making this the default!

Tarrasch commented:
Thanks for keeping us updated. Really looking forward to this! :)

stephenpascoe commented:
Visualiser update has landed.  Everything @daveFNbuck mentioned has been addressed except restyling the worker tab.

Supporting task numbers greater than max-shown-tasks was quite complex.  My solution is to add a "Filter on Server" checkbox which will switch on calling the server API when filtering.  It is probably still a  bit confusing.

The default visualiser is now SVG but the browser will remember which one you chose for next time (via cookie).

Tarrasch commented:
So awesome!!!!! :D

Tarrasch commented:
Want to take another look @daveFNbuck? I'm going to try it out now :)

Tarrasch commented:
Ok, I'm building your branch rebased to make sure tests still pass.

I found one small bug, but it's not blocking a merge.

 1. Go to a dependency graph view (with either visualiser)
 2. Click "show upstream dependencies"
 3. It works as expected
 4. Click reload (F5) in browser
 5. Checkbox for "show upstream dependencies" is still checked. But it shows the downstream dependencies.

Apart from that nitpick:

It looks fantastic! It's snappy, easy to understand and the buttons makes sense. Let's merge? Any final thoughts @daveFNbuck?

stephenpascoe commented:
Yep, I see the reload bug on Firefox but not on Chrome. I've also noticed the visualiser isn't working on Safari.  Generally I've developed on Chrome and tested a little on Firefox.

Tarrasch commented:
Ah, thanks for looking it up and taking care of it. :)

daveFNbuck commented:
I'll definitely try this out today. Thanks for getting all this done so
quickly!

On Fri, Sep 11, 2015 at 8:06 AM, Arash Rouhani <notifications@github.com>
wrote:

> Ah, thanks for looking it up and taking care of it. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/luigi/pull/1086#issuecomment-139570132>.
>

daveFNbuck commented on luigi/static/visualiser/js/visualiserApp.js:
```diff
@@ -50,35 +80,145 @@ function visualiserApp(luigi) {
 
     function processWorker(worker) {
-        worker.tasks = worker.running.map($.proxy(taskToDisplayTask, null, false));
+        worker.tasks = worker.running.map($.proxy(taskToDisplayTask, false, null));
```
This call does not work because $.proxy is weird. Luckily, we don't need the proxy function now that you've changed the order of the arguments to taskToDisplayTask and made the second one optional. Just do

    worker.tasks = worker.running.map(taskToDisplayTask);

and that'll fix it.

daveFNbuck commented:
It works pretty well for me now, at least on Chrome. I think this is good to merge except for the bug I mentioned in my diff comment a few minutes ago. I've got a few minor issues, but these can be dealt with separately after this gets merged:

* All the filters and warnings/info at the top of the page should be hidden like the task families when on the graph or workers page
* The too many X tasks to display warning should go away when my filters don't include type X
* Server-side filtering can be a little slow sometimes and there's no indication that it's working on something. I can work a bit on speeding it up, but it would be very helpful to have a spinner or something while we're waiting on the response.
* I love that with server-side filtering, the task families and numbers under the statuses get updated. It would be great if this worked with client-side filtering too
* It would be nice to remember whether I had server-side filtering checked
* It would be nice to remember my filters for a short period so I could refresh the data without having to reselect the filters

daveFNbuck commented:
Actually, it might be even nicer for the filters to apply to the graphs too. I often want to see a graph without all the DONE tasks to get a better idea of what still needs to happen.

daveFNbuck commented:
I found one more bug that I think needs fixing. The error messages contain all the json escapes, so they have \n instead of newlines and are generally peppered with slashes. This results in one extremely long line, and it's not even possible to scroll. It'd be much better if these were unescaped.

Tarrasch commented:
I recall seeing proper newlines. Maybe I misremember though.

I really like the idea of filtering on graphs too. But that doesn't need to be in this PR of course. :)

daveFNbuck commented:
There were proper newlines in the previous version, but I'm not seeing them today.

stephenpascoe commented:
I am seeing this bug once I rebase but not from where I was initially branched.  Something else must have changed.  I'll take a look.

stephenpascoe commented:
I've rebased and added 1 commit which fixes the newline issue.  It tweaks commit https://github.com/spotify/luigi/commit/f5267517.

Re @daveFNbuck's other points.  Once this is merged I'd be keen to refactor the javascript.  It is an amalgamation of the original, the D3 code and my changes; there is plenty of dead code; the callbacks are messy.  These other features would be easier to do after that.  I will need to focus on other things for a few weeks though.

daveFNbuck commented:
Looks good to me if you just change line 221 in visualiserApp.js to

    worker.tasks = worker.running.map(taskToDisplayTask);

and fix whatever caused travis to break

Tarrasch commented:
Cool, this is soon reaching completion it seems. :)

Tarrasch commented:
Cool.  Is there anything left to be done for this @stephenpascoe?

stephenpascoe commented:
IMO It's done.

Tarrasch commented:
Huzzah!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Awesome work @stephenpascoe!!

ulzha commented:
Well done! Super clear legend and all.

erikbern commented:
![champagne](https://cloud.githubusercontent.com/assets/1027979/9936998/a6c7d208-5d2c-11e5-8d7a-417b70b7cc57.gif)

stephenpascoe commented:
Thanks all, and thanks to @beeva-luisgonzalez for the D3 code.
